### PR TITLE
Yearn linear pool and supporting utilities

### PR DESCRIPTION
**Pull request by danielmkm**
Implementation of the `YearnLinearPool`, already deployed to the Fantom Opera network on Beethoven X, modified to support the new Rebalancer model. The YearnLinearPool supports the latest iteration of Yearn's multi-strat single asset vaults. Common instances of this vault include yvUSDC, yvDAi, yvUSDT, yvBTC, yvETH, etc. The implementation leverages the vault exposed price per share (pps) when inferring the `wrappedTokenRate`.

This PR contains the following:
- YearnLinearPool
- YearnLinearPoolFactory
- YearnLinearPoolRebalancer
- YearnWrapping (BatchRelayer extension)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203292188855174

### CODE REVIEW

File: pkg/pool-linear/contracts/yearn/YearnLinearPoolRebalancer.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        IERC20 wrappedToken = ILinearPool(ILastCreatedPoolFactory(msg.sender).getLastCreatedPool()).getWrappedToken();
+
+        _divisor = 10**IYearnTokenVault(address(wrappedToken)).decimals();
+    }
+
+    function _wrapTokens(uint256 amount) internal override {
+        // Depositing from underlying (i.e. DAI, USDC, etc. instead of yvDAI or yvUSDC). Before we can
+        // deposit however, we need to approve the wrapper (yearn vault) in the underlying token.
+        _mainToken.approve(address(_wrappedToken), amount);
+        IYearnTokenVault(address(_wrappedToken)).deposit(amount, address(this));
```
Comment (joaobrunoah) [inline pkg/pool-linear/contracts/yearn/YearnLinearPoolRebalancer.sol:44]: This line is failing in forked tests. Do you need to inform the recipient's address in the deposit function?

**Review by joaobrunoah — CHANGES_REQUESTED** [review]

File: pkg/pool-linear/contracts/yearn/YearnLinearPoolRebalancer.sol (hunk 1 above)
Comment (danielmkm) [reply pkg/pool-linear/contracts/yearn/YearnLinearPoolRebalancer.sol:46]: Hey joão! We revised the PR to use the YearnShareValueHelper and added fork tests that pass.

**Comment by danielmkm** [comment]
We updated this PR to leverage the ShareValueHelper implemented by @wavey0x (yearn dev).
https://github.com/wavey0x/YearnSharePriceConverter/blob/master/contracts/Helper.sol

We also included functional fork tests.

The ShareValueHelper provides a more precise wrappedTokenRate than the exposed pricePerShare (pps) since the pps exposes only as many decimals as the underlying token (ie: USDC = 6), but it actually stores higher precision internally.

An instance of IYearnShareValueHelper is passed into the YearnLinearPoolFactory as a constructor argument:

https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/pool-linear/contracts/yearn/YearnLinearPoolFactory.sol#L37-L42

https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/pool-linear/contracts/yearn/YearnLinearPoolFactory.sol#L46-L54


Then on any call to `factory.create`, the instance of IYearnShareValueHelper is passed to the YearnLinearPool and YearnLinearPoolRebalancer

https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/pool-linear/contracts/yearn/YearnLinearPoolFactory.sol#L96-L99

https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/pool-linear/contracts/yearn/YearnLinearPoolFactory.sol#L104-L117


https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/deployments/tasks/20221114-yearn-linear-pool/input.ts#L14-L51


The YearnLinearPool leverages IYearnShareValueHelper.sharesToAmount() when determining the _getWrappedTokenRate

https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/pool-linear/contracts/yearn/YearnLinearPool.sol#L75-L79


The YearnLinearPoolRebalancer leverages IYearnShareValueHelper.sharesToAmount() when determining the _getRequiredTokensToWrap

https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/pool-linear/contracts/yearn/YearnLinearPoolRebalancer.sol#L55-L61

For the moment we've defined the instance of IYearnShareValueHelper in the input file. The yearn team will deploy official version that we will update prior to this getting merged. Alternatively we can bring the code into the mono-repo if its desired, as it has no dependencies to external libraries (other then the YearnTokenVault).
https://github.com/balancer-labs/balancer-v2-monorepo/blob/74afb21d69e7dd37cadfd528692d79615d0c0478/pkg/deployments/tasks/20221114-yearn-linear-pool/input.ts#L14-L51

**Comment by mkflow27** [comment]
Hey @danielmkm ,

future LinearPool integrations will live in a separate Repo for the Balancer protocol. This repo will be maintained by Orb collective - a Balancer Serivce Provider (https://github.com/orbcollective). We are in the process of setting this repo up. I wanted to already reach out and inform you about it. Once it is set up, I will reach out with more details

**Comment by EndymionJkb** [comment]
Overtaken by events - moved to Orb repo

**Pull request closed without merging**